compiler/il/OMRBlock.cpp

```diff
@@ -805,6 +807,687 @@ OMR::Block::splitWithGivenMethodSymbol(TR::ResolvedMethodSymbol *methodSymbol, T
    return self()->split(startOfNewBlock, cfg, fixupCommoning, copyExceptionSuccessors, methodSymbol);
 }
 
+// Required data structures for analysis in SplitPostGRA
+typedef TR::typed_allocator<std::pair<TR::Node *, std::pair<int32_t, TR::Node *> >, TR::Region&> NodeTableAllocator; 
+typedef std::less< TR::Node *> NodeTableComparator; 
+typedef std::map<TR::Node *, std::pair<int32_t, TR::Node *>, NodeTableComparator, NodeTableAllocator> NodeTable;
+
+typedef TR::typed_allocator<std::pair <int32_t, TR::Node *>, TR::Region&> StoreRegNodeTableAllocator;
+typedef std::map<TR::Node *, TR::Node *, NodeTableComparator, StoreRegNodeTableAllocator> StoreRegNodeTable;
+
+/*
+ * Recursive call to replace node with replacement node underneath a target node.
+ *
+ * @param A node to be replaced
+ * @param A table containing information about original nodes to be replaced with corresponding replacement node
+ * @param checklist Checklist of nodes that have already been searched.
+ */
+static void replaceNodesInSubtree(TR::Node *node, NodeTable *nodeInfo, TR::NodeChecklist &checklist)
+   {
+   if (checklist.contains(node))
+      return;
+
+   checklist.add(node);
+
+   for (int i = 0; i < node->getNumChildren(); ++i)
+      {
+      TR::Node *child = node->getChild(i);
+      auto entry = nodeInfo->find(child);
+      if (entry != nodeInfo->end())
+         {
+         node->setAndIncChild(i, entry->second.second);
+         child->decReferenceCount();
+         }
+      else
+         {
+         replaceNodesInSubtree(child, nodeInfo, checklist);
+         }
+      }
+   }
+
+/*
+ * Search for the nodes to be uncommoned from the split point till end of extended basic block
+ *
+ * @param Compilation object
+ * @param start tree top from where it needs to start searching
+ * @param table containing information about original nodes to be replaced with corresponding replacement node 
+ */
+static void replaceNodesInTrees(TR::Compilation * comp, TR::TreeTop *start, NodeTable *nodeInfo)
+   {
+   TR::NodeChecklist checklist(comp);
+   while (start)
+      {
+      if (start->getNode()->getOpCodeValue() == TR::BBStart && !start->getNode()->getBlock()->isExtensionOfPreviousBlock())
+         break;
+      replaceNodesInSubtree(start->getNode(), nodeInfo, checklist);
+      start = start->getNextTreeTop();
+      }
+   }
+
+/*
+ * Analyzes the node and list of unavailable registers to find a register pair which can be used to store the node
+ *
```

Comment:
What is done with the nodes that are commoned across a split point ? Are they stored into "available" global registers and carried across the new CFG edges that have been created ?

Reply:
Yes the overall approach is to figure out the set of unused global registers at the split point. Any values commoned across the split will be carried across the created edge using global registers. If we run out of global registers we revert to using the stack by creating temps in the traditional way.

The comment above the main API further down provides a good high-level overview.

Reply:
Thanks @andrewcraik  for the explanation. Yes this API will try to use the `available` global register to common nodes across the new CFG edges and will use stack as last resource if it can not find the registers.